### PR TITLE
Implement the dagster-openai integration library

### PR DESCRIPTION
**Pull request by maximearmstrong**
## Summary & Motivation

This PR adds a new `dagster-openai` library to our set of libraries.

The main goal of this library is to log the Open AI API usage in the metadata. To do so, we need to wrap the methods called through the client, get the results and update the metadata.

Initial code snippets was hardcoding 3 methods, but we want to give the user some flexibility.

Constraints:
- Results must be captured at the method level - the data we seek is included in the OpenAI API response. The results can't be captured at the client level, at teardown for instance.
- Not all the methods existing in the OpenAI library should be wrap (private methods, etc.)
- Methods are overloaded in the API Resource classes, so wrapping the methods should be done on the instance.

**Solution**

Implement `OpenAIResource.get_client`, `OpenAIResource.get_client_for_asset` and the function wrapper `with_usage_metadata`. 

By default, for assets, the methods for the 3 main API Endpoint classes, `Completions`, `Chat` and `Embeddings`, are wrapped when instantiating the client - wrapping the methods allows to log the usage metadata provided in an OpenAI Completion response. If another endpoint should be wrapped, a user can use `with_usage_metadata` to it and log the metadata.

`OpenAIResource.get_client` can be used for assets and ops, but the metadata will not be logged for ops. `OpenAIResource.get_client_for_asset` can only be used with assets and the metadata will be logged.

## TO-DOs
- [x] implement the resource
- [x] add docstrings
- [x] implement tests

## How I Tested These Changes
Local implementation
BK
Dogfood in Purina with a toy example

### CODE REVIEW

**Comment by maximearmstrong** [comment]
Current dependencies on/for this PR:
* `master`
  * **PR #19697** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/19697?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/19697?utm_source=stack-comment).

File: python_modules/libraries/dagster-openai/dagster_openai/resources.py
Comment (yuhan) [inline python_modules/libraries/dagster-openai/dagster_openai/resources.py:111]: nit: we can probably call the add_to_asset_metadata once since we will have all the dict value ready here [1]

File: python_modules/libraries/dagster-openai/dagster_openai/resources.py
Comment (yuhan) [inline python_modules/libraries/dagster-openai/dagster_openai/resources.py:56]: re: [1], then you probably change it so it's update dict instead of add single value to a metadata key

File: python_modules/libraries/dagster-openai/dagster_openai/resources.py
Comment (yuhan) [inline python_modules/libraries/dagster-openai/dagster_openai/resources.py:89]: worth adding comment here to explain different behaviors

File: python_modules/libraries/dagster-openai/dagster_openai/resources.py
Comment (yuhan) [inline python_modules/libraries/dagster-openai/dagster_openai/resources.py:243]: i was imagining this to be an internal implementation to start with.

[2] seems a bit heavy IMO - maybe we could stake a separate PR for this pattern. the MVP can just not allow api_resources - it's easier and better to restrict more and open it up later.

File: python_modules/libraries/dagster-openai/dagster_openai/resources.py
Comment (yuhan) [inline python_modules/libraries/dagster-openai/dagster_openai/resources.py:218]: [2]

**Review by yuhan — COMMENTED** [review]
my overall reaction is we can limit the MVP to only enable a subset of APIs and allow for more in a later stack

**Comment by yuhan** [comment]
have you tested this in a toy dagster project and how does it look like in 1) asset metadata which is available across oss and cloud, 2) cloud insights?

File: python_modules/libraries/dagster-openai/dagster_openai/resources.py
Comment (maximearmstrong) [reply python_modules/libraries/dagster-openai/dagster_openai/resources.py:27]: We could start with the 3 main ones, `Chat`, `Completions` and `Embeddings` and add more from there?

File: python_modules/libraries/dagster-openai/dagster_openai/resources.py
Comment (maximearmstrong) [reply python_modules/libraries/dagster-openai/dagster_openai/resources.py:243]: It makes sense, I can remove the option.

**Comment by maximearmstrong** [comment]
> have you tested this in a toy dagster project and how does that look like in 1) asset metadata which is available across oss and cloud, 2) cloud insights?

It was fully tested locally with a toy dagster project. Next step was cloud, I wanted to make sure we were on the right track before completing all tests.

File: python_modules/libraries/dagster-openai/dagster_openai/resources.py
Comment (maximearmstrong) [reply python_modules/libraries/dagster-openai/dagster_openai/resources.py:13]: > Naming quibble - since this name overloads the Dagster "Resource" noun, it could be pretty confusing to users (they construct an OpenAI Resource & pass it API resources). Are these called resources in OpenAI docs?

I agree, but indeed, they are called resources in the OpenAI library we are wrapping. I also thought of ApiEndpointClassesEnum to make it obvious that it is not related to a dagster resource. I can make the change.

File: python_modules/libraries/dagster-openai/dagster_openai/resources.py
Comment (maximearmstrong) [reply python_modules/libraries/dagster-openai/dagster_openai/resources.py:187]: As it is now, it would support assets only, but I agree that it would be useful. 

@Yuhan, WDYT, should we support op context for the MVP or later?

File: python_modules/libraries/dagster-openai/dagster_openai/resources.py
Comment (yuhan) [reply python_modules/libraries/dagster-openai/dagster_openai/resources.py:13]: `ApiEndpointClassesEnum` sounds not bad to me - i'd lean towards very descriptive to start with.

**Comment by maximearmstrong** [comment]
@yuhan @benpankow I updated the code in [ebe9367](https://github.com/dagster-io/dagster/pull/19697/commits/ebe936761952ee7a6aa74d00599f5c7503e74b9e) to match the first reviews. 

User can now use

- `OpenAIResource.get_client(context)` for `AssetExecutionContext` and `OpExecutionContext`
  - We don't wrap the methods for op context because we don't capture metadata in that scenario.
- `with_usage_metadata()` to wrap any endpoint they want.
  - we support only `Chat`, `Completions` and `Embeddings` endpoint classes by wrapping them by default in `get_client(context)` when context is instance of `AssetExecutionContext`.
  
The code was tested in a toy repo locally and in local cloud deployment. 

Missing:
- Examples in the docstrings
- Unit tests

File: python_modules/libraries/dagster-openai/dagster_openai/resources.py
Comment (yuhan) [inline python_modules/libraries/dagster-openai/dagster_openai/resources.py:176]: note on unit tests, since you have output_name here: worth testing the behavior in the following cases:
1. multi_assets (https://docs.dagster.io/concepts/assets/multi-assets)
2. assets with partitions (https://docs.dagster.io/concepts/partitions-schedules-sensors/partitioning-assets)
3. graph-backed assets (https://docs.dagster.io/concepts/assets/graph-backed-assets) - i doubt this will behave any different than a regular asset but worth throwing a test case for complexity

**Review by yuhan — COMMENTED** [review]
looking close! look forward to the complete docstring and test cases!

File: python_modules/libraries/dagster-openai/dagster_openai/resources.py
Comment (maximearmstrong) [reply python_modules/libraries/dagster-openai/dagster_openai/resources.py]: All done in [dagster-openai/dagster_openai_tests/test_resources.py](https://github.com/dagster-io/dagster/pull/19697/files#diff-db7e37302fefbb53ccf043549f955c4aea4c4ecc275257ee6dc163c88059a5d1)

File: python_modules/libraries/dagster-openai/dagster_openai/resources.py
Comment (maximearmstrong) [reply python_modules/libraries/dagster-openai/dagster_openai/resources.py:230]: Done in [bbe2b6e](https://github.com/dagster-io/dagster/pull/19697/commits/bbe2b6e7b5f03d6a92f992d899f6aba1615b6f1f)

File: python_modules/libraries/dagster-openai/dagster_openai/resources.py
Comment (yuhan) [inline python_modules/libraries/dagster-openai/dagster_openai/resources.py:216]: per internal convo, might worth switching to `get_client` and `get_client_for_asset` to keep the consistency with InsightsSnowflakeResource

File: python_modules/libraries/dagster-openai/dagster_openai_tests/test_resources.py
Comment (yuhan) [inline python_modules/libraries/dagster-openai/dagster_openai_tests/test_resources.py:146]: lets update it to using `specs` which is a more blessed pattern.

File: python_modules/libraries/dagster-openai/dagster_openai_tests/test_resources.py
Comment (yuhan) [inline python_modules/libraries/dagster-openai/dagster_openai_tests/test_resources.py:196]: same as [2]

File: python_modules/libraries/dagster-openai/dagster_openai_tests/test_resources.py
Comment (maximearmstrong) [reply python_modules/libraries/dagster-openai/dagster_openai_tests/test_resources.py]: Done in [d3581d5](https://github.com/dagster-io/dagster/pull/19697/commits/d3581d571913914f9a18087bb3268503c53894ba)

File: python_modules/libraries/dagster-openai/dagster_openai_tests/test_resources.py
Comment (maximearmstrong) [reply python_modules/libraries/dagster-openai/dagster_openai_tests/test_resources.py]: Done in [d3581d5](https://github.com/dagster-io/dagster/pull/19697/commits/d3581d571913914f9a18087bb3268503c53894ba)

File: python_modules/libraries/dagster-openai/dagster_openai_tests/test_resources.py
Comment (maximearmstrong) [reply python_modules/libraries/dagster-openai/dagster_openai_tests/test_resources.py]: Done in [d3581d5](https://github.com/dagster-io/dagster/pull/19697/commits/d3581d571913914f9a18087bb3268503c53894ba)

File: python_modules/libraries/dagster-openai/dagster_openai/resources.py
Comment (maximearmstrong) [reply python_modules/libraries/dagster-openai/dagster_openai/resources.py]: Done in [ec36d2c](https://github.com/dagster-io/dagster/pull/19697/commits/ec36d2c3570230d6942e927c6ee95f88babb41f2)

File: python_modules/libraries/dagster-openai/dagster_openai/resources.py
Comment (yuhan) [inline python_modules/libraries/dagster-openai/dagster_openai/resources.py:215]: probably need a `@public` decorator here to make it show up in doc string. we can add and test that in the docs PR tho.

File: python_modules/libraries/dagster-openai/dagster_openai/resources.py
Comment (yuhan) [inline python_modules/libraries/dagster-openai/dagster_openai/resources.py:276]: `@public`

**Review by yuhan — APPROVED** [review]
Let's merge this in! And iterate if anything else comes up as we build some examples using it.

Before landing, please make sure we aren't releasing it this week. See how to whitelist in release pipeline here: https://dagsterlabs.slack.com/archives/C03A0D72A6T/p1708645224684229

**Pull request merged**
